[WIP] Provide CMake build system

robertodr opened the pull request:
This PR introduces a build system based on CMake, version 3.3 and later. The CMake build system coexists with the one already in place, but should make it possible to streamline the compilation of the code. To try it out:
1. Generate the build system:
    ```
    ./mkconfig.py
   ```
   This is merely a Python front-end to CMake, it is equivalent to running:
   ```
   cmake -H. -Bbuild
   ```
   and will generate the build system (_i.e._ the `Makefile`-s) in a directory called `build`
2. Build the code:
    ```
    cd build
    make -jN
    ```

## TODOs
The PR is still classed as work-in-progress. These are the issues I aim at addressing before it's ready to go:
- [ ] Wrap the test suite with CTest. Once the code is built, users should be able to run:
  ```
  ctest
  ```
  to run the test suite. This is merely a wrapper to `testcode.py` and is not hard to implement. I have however made some modifications to `lib/local/environment_report.F90` that seem to disrupt the data extraction executed _via_ `pyhande`. I am investigating this.
- [ ] I have tested that everything **builds** on my local machine. I need to get this tested on a real cluster to really confirm that it streamlines the build procedure.
- [ ] Setting compiler flags can be done passing options to `mkconfig.py` or CMake directly, but the defaults are probably not very good neither in Debug nor in Release configuration. I plan to change that and expose a clear mechanism for developers to set default flags for the different configurations.

jsspencer commented:
This looks very nice. I have several questions and issues though. I would also like to spend some time on a more in-depth review.

1. What is your local machine? I guess NixOS from the addition of default.nix. I am in two minds about this -- NixOS is pretty cool (and I would use it if I didn't like arch so much) but I've always been reluctant to add config files for personal setups (e.g. IDE related files), at least to the top-level. Would be happier if it was a dotfile or in a subdirectory.
1. Personal bias: I like make and have a grudging acceptance that cmake is very useful. I also have painful experience that getting code to compile on weird HPC setups can require lots of tweaks to the build and this is much easier to do with a simple makefile than the one generated by cmake (which I always find quite opaque). Would like to keep compilation via the existing makefile for this reason.
1. It's not immediately obvious how to change the link line (e.g. add libraries) from mkconfig.py.
1. Which brings me onto my main point -- I finally got hande to compile by running NixOS inside docker (I have too many operating systems in my life as it is). There seem to be problems on both OS X and Arch (haven't tried elsewhere). On OS X, mkconfig.py finds BLAS and LAPACK but then this isn't added to the link line:
```
$ mkconfig.py
[snip]
-- Searching for BLAS using search order MKL;ESSL;OPENBLAS;ATLAS;ACML;SYSTEM_NATIVE
-- Setting -DHAVE_SYSTEM_NATIVE_BLAS
-- Found BLAS: SYSTEM_NATIVE (/usr/lib/libblas.dylib)
-- Searching for LAPACK using search order MKL;ESSL;OPENBLAS;ATLAS;ACML;SYSTEM_NATIVE
-- Setting -DHAVE_SYSTEM_NATIVE_LAPACK
-- Found LAPACK: SYSTEM_NATIVE (/usr/lib/liblapack.dylib)
-- MATH_LIB_SEARCH_ORDER set to MKL;ESSL;OPENBLAS;ATLAS;ACML;SYSTEM_NATIVE
[snip]
$ make -C build VERBOSE=1
[snip]
/usr/local/bin/gfortran -g -dynamiclib -Wl,-headerpad_max_install_names -compatibility_version 1.0.0 -o ../lib/libhande.1.dylib -install_name @rpath/libhande.1.dylib [list of object files] ../lib/libaotus.a /usr/local/lib/liblua.5.3.dylib
Undefined symbols for architecture x86_64:
  "_cheev_", referenced from:
      ___linalg_MOD_cheev_f90 in linalg.F90.o
     (maybe you meant: ___linalg_MOD_cheev_f90)
  "_dgeev_", referenced from:
      ___linalg_MOD_dgeev_f90 in linalg.F90.o
     (maybe you meant: ___linalg_MOD_dgeev_f90)
  "_dsyev_", referenced from:
      ___linalg_MOD_dsyev_f90 in linalg.F90.o
     (maybe you meant: ___linalg_MOD_dsyev_f90)
  "_sgeev_", referenced from:
      ___linalg_MOD_sgeev_f90 in linalg.F90.o
     (maybe you meant: ___linalg_MOD_sgeev_f90)
  "_ssyev_", referenced from:
      ___linalg_MOD_ssyev_f90 in linalg.F90.o
     (maybe you meant: ___linalg_MOD_ssyev_f90)
  "_zheev_", referenced from:
      ___linalg_MOD_zheev_f90 in linalg.F90.o
     (maybe you meant: ___linalg_MOD_zheev_f90)
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
```
The maths libraries are not added to the link line. Doing
```
$ LDFLAGS='-llapack -lblas' ./mkconfig.py && make -C build
```
worked however.
Even on NixOS it wasn't straightforward -- I couldn't find the atlasWithLapack package on https://nixos.org/nixos/packages.html and had to change this to liblapack in default.nix to get all the dependencies. Is atlasWithLapack a local package? If so, it would be helpful for default.nix to work for others as well. (It also includes some dependencies which aren't strictly required by hande, such as sympy.)
5. Finally build instructions need to be rewritten but I appreciate this is a WIP and they will be left to the end.
5. I am not familiar with autocmake but it looks like it generates a set of files which are then added to the repository. Is this correct? Is there a license associated with them? (There's nothing indicating this in the files under cmake/ but we need to be careful about adding code to the repository.) At the very least, a README under cmake/ would be helpful.

I think this will be a really valuable contribution -- there have been frequent issues with new people compiling hande and this will go a long way to helping resolve them. Thanks!

jsspencer commented on lib/local/print_info.c:
```diff
@@ -0,0 +1,79 @@
+/*
+ * Print the configuration/build info generated at build time by the executable.
+ * We do this from C and not from Fortran because Fortran has trouble with long
+ * lines.
```
Source/example? Fortran string handling has always been incredibly ugly though...

robertodr replied:
No specific source, just personal experience with Fortran string handling back and forth with C/C++.

jsspencer replied:
Well, I don't have any real objections given the amount of mixed code we're already doing and string formatting in fortran is horrible. (I do slightly regret picking Fortran for HANDE...)

I saw some weirdness with buffer flushing but I think that's resolved now.

jsspencer commented on lib/local/environment_report.F90:
```diff
@@ -126,73 +80,13 @@ end function getcwd_c
-#endif
-        write (io_unit,'(5X,"DET_SIZE = ",i2,".")') bit_size(0_i0)
-        write (io_unit,'(5X,"POP_SIZE = ",i2,".")') bit_size(0_int_p)
+        write (io_unit,'("Calculation UUID:",1X,a36)') new_line('a')//GLOBAL_UUID
```
Why the change for the output format of the UUID line with a new line? Also (for consistency) prefer using / in for format string instead of new_line.

robertodr replied:
I thought it would fit better with the rest of the printing for that section. But on second thought it looks uglier, so I'll revert it.

robertodr commented:
Thanks for your feedback @jsspencer! It's highly appreciated:
1. Yes, I am on NixOS. I use `default.nix` to start a `nix-shell` with the right library dependencies. I am still on Nixos 17.03 and I've just realized that `atlasWithLapack` is not under the list of packages you linked in your comment. I've corrected that, plus I've renamed the file `.default.nix`. If you use [`direnv`](https://direnv.net/) the `.envrc` will load the `nix-shell` as soon as you jump into the directory. I use `tmux` a lot and this makes it very easy to have a coherent set up across all windows and panes.
The `nix-shell` I run is not pure. I'm probably picking up some dependencies from my global configuration and that might be manifesting with the failures you're reporting.
2. I definitely concur with this! Not because I like `Makefile`-s (I don't know anything about them, having "grown up" with CMake), but because the current set up has been honed for a long time. It would be madness to throw it away, especially on those machines where the right configuration has been found.
3. I opened an issue on my fork to include that. It's something I overlooked.
4. I'll investigate both problems:
  a. The BLAS/LAPACK issue is a bit weird. I seem to recall that the CMake functions for math detection were acting up a bit on Mac OS X.
  b. Yes, I need to `-fPIC` the whole build. I've noticed that some versions of CMake do it by default, some others don't (possibly also depending on your compiler) This is part of the larger issue of setting compiler flags consistently throughout the project, checking if they're available for the given compiler. There's now an issue open on my fork. This is a general problem and I'm working on a solution to include in Autocmake.
5. [Autocmake](http://autocmake.readthedocs.io/en/latest/) is a CMake plugin composer, where the plugins are snippets of CMake code that take care of one task (like finding math libraries). Autocmake is [BSD-3-Clause](https://github.com/coderefinery/autocmake), which is [remixable](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses) with HANDE's LGPL-2.1
However, you're right, it is not clear from the source code of the plugins I did not write what the license is. @bast is the main maintainer of the Autocmake project and coderefinery/autocmake#211

bast commented:
Thanks for the good feedback on making the license terms more visible. I will take care of this.

jsspencer commented:
1. Thanks for the introduction to nix. I've been watching it for some time and quite like aspects of it now I've actually used it!
2. Great. cmake is great but sometimes nothing beats manual hacking...
3. I saw, thanks!
4. a. Thanks. Let me know if there's testing that I can do if you don't have access to OS X.
    b. Slightly curious -- does -fPIC have an impact on performance? I think it can if it hinders interprocedural optimsations (which can help quite a bit). It's been several years since I looked at this though.
5/-. Suggest documentation once everything else is settled.
6/5. This is great. Please include an update to get this. Thanks @bast.

robertodr commented:
The `nix-shell` is now pinned to a specific commit on the current stable channel of NixOS. Trick learnt here: http://techblog.holidaycheck.com/post/2017/11/14/nix-in-practice-providing-dependencies There are also a couple of overlays defined to get the correct ATLAS and HDF5. HDF5 is still without Fortran2003 bindings, I need to send a PR to the nixpkgs repo for that...
The shell is now created persistently, meaning that a garbage collection will not clean it up. This is very handy, since I spent the whole of last night compiling ATLAS from scratch.

jsspencer commented:
That's nice! I'm a little sad that the syntax for doing that in Nix looks rather convoluted.

I share your atlas pain (though my use cases are now so simple I've given up and just use netlib implementations if I don't have MKL available). It is a shame that we need Fortran2003 support for HDF5 (which makes things much, much easier) and even more that this is still not enabled by default. It's almost 2018!

robertodr commented:
NixOS/nixpkgs#31781 is the PR I made for HDF5. Let's see if it makes to a stable channel. It actually turns out that with HDF5 1.10.1 `--enable-fortran` gives the Fortran2003 bindings. I've been working on some other stuff this week, but I'll jump back on this on Monday. It would be awesome to have this merged by the end of the year :smile:

jsspencer commented:
Thanks! We also have a bunch of code to push which we're tidying up but I think this should be mergeable by the end of the year. :D

robertodr commented:
I've restored CI on Travis. It's now testing Linux and OS X with Python 2.7 and 3.5 (4 builds) It's not testing the CMake build system yet, though. I will most likely need to cut back on the build matrix to keep the CI running time acceptable.

robertodr commented:
At long last an update:
1. I can build the executable on Travis and that should also fix your `-fPIC` problems @jsspencer[^1]
However, tests are still failing. It seems due to faulty data extraction:
```
Traceback (most recent call last):

  File "/home/travis/Deps/testcode/bin/testcode.py", line 804, in <module>

    sys.exit(main(sys.argv[1:]))

  File "/home/travis/Deps/testcode/bin/testcode.py", line 782, in main

    run_tests(tests, verbose, options.queue_system, options.tot_nprocs)

  File "/home/travis/Deps/testcode/bin/testcode.py", line 392, in run_tests

    test.run_test(verbose, cluster_queue, os.getcwd())

  File "/home/travis/Deps/testcode/bin/../lib/testcode2/__init__.py", line 331, in run_test

    self.verify_job(test_input, test_arg, verbose, rundir)

  File "/home/travis/Deps/testcode/bin/../lib/testcode2/dir_lock.py", line 29, in decorated_func

    return func(*args, **kwargs)

  File "/home/travis/Deps/testcode/bin/../lib/testcode2/dir_lock.py", line 59, in decorated_func

    val = func(*args, **kwargs)

  File "/home/travis/Deps/testcode/bin/../lib/testcode2/__init__.py", line 465, in _verify_job

    verbose)

  File "/home/travis/Deps/testcode/bin/../lib/testcode2/__init__.py", line 579, in extract_data

    outputs = [tp_ptr.extract_fn(dfile) for dfile in data_files]

  File "../tools/pyhande/pyhande/testcode.py", line 84, in testcode_data

    for (key, dat) in extract_test_data(fname, False).items():

  File "../tools/pyhande/pyhande/testcode.py", line 24, in extract_test_data

    hande_out = pyhande.extract.extract_data(fname)

  File "../tools/pyhande/pyhande/extract.py", line 142, in extract_data

    (metadata, data, comment_data) = _extract_mc_calc(f, calc_type)

  File "../tools/pyhande/pyhande/extract.py", line 280, in _extract_mc_calc

    data['iterations'] = pd.to_numeric(data['iterations'])

  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/pandas/core/tools/numeric.py", line 133, in to_numeric

    coerce_numeric=coerce_numeric)

  File "pandas/_libs/src/inference.pyx", line 1158, in pandas._libs.lib.maybe_convert_numeric

ValueError: Unable to parse string "Basis" at position 0
```
2. I'm working on math detection. Hopefully something more concrete next week.

[^1] Lua built from source only provides the static, non `fPIC`-ed library, which makes trouble when trying to link into a shared library, like I was trying to. I never noticed the problem on my machines because most distros package both static and shared Lua and CMake prefers shared over static.

robertodr commented:
Closing and reopening as #3
